File: apps/web/src/types/methods.ts
export type MethodCategory = "quantitative" | "configurational";
export type ResultBlockType = "summary" | "table" | "chart" | "text" | "download";

export type MethodFieldSpec = {
  name: string;
  label: string;
  type: string;
  required: boolean;
  role: string;
  description: string;
};

export type MethodFieldSpecDto = MethodFieldSpec & {
  allow_missing: boolean;
  constraints: {
    min: number | null;
    max: number | null;
    allowed_values: string[] | null;
  } | null;
};

export type MethodInputSpec = {
  acceptedFileTypes: Array<"csv" | "xlsx">;
  minSampleSize?: number;
  fields: MethodFieldSpec[];
};

export type MethodInputSpecDto = {
  spec_version: string;
  accepted_file_types: Array<"csv" | "xlsx">;
  min_sample_size: number | null;
  fields: MethodFieldSpecDto[];
};

export type MethodResultBlockDto = {
  key: string;
  type: ResultBlockType;
  title: string;
  required: boolean;
};

export type MethodOutputSpec = {
  summary: string;
  resultBlocks: string[];
};

export type MethodOutputSpecDto = {
  summary: string;
  result_blocks: MethodResultBlockDto[];
};

export type MethodResource = {
  label: string;
  url: string | null;
  status: "placeholder";
};

export type MethodListItemDto = {
  id: string;
  slug: string;
  name: string;
  category: MethodCategory;
  description: string;
  applicable_scenarios: string[];
  data_type_label: string;
  analysis_difficulty: "低" | "中" | "高";
  sample_requirement: string;
  is_recommended: boolean;
};

export type MethodDetailDto = MethodListItemDto & {
  input_spec_json: MethodInputSpecDto;
  output_spec_json: MethodOutputSpecDto;
  template_file_url: string | null;
  sample_file_url: string | null;
  prerequisites: string[];
  common_errors: string[];
  is_active: boolean;
  display_order: number;
  created_at: string;
  updated_at: string;
};

export type ApiSuccessResponse<T> = {
  success: true;
  message: string;
  data: T;
};

export type ApiErrorResponse = {
  success: false;
  message: string;
  error_code: string;
  details: Record<string, unknown>;
};

export type Method = {
  id: string;
  slug: string;
  name: string;
  category: MethodCategory;
  description: string;
  applicableScenarios: string[];
  dataTypeLabel: string;
  analysisDifficulty: "低" | "中" | "高";
  sampleRequirement: string;
  prerequisites: string[];
  inputSpec: MethodInputSpec;
  outputSpec: MethodOutputSpec;
  templateResource: MethodResource;
  sampleResource: MethodResource;
  commonErrors: string[];
  isRecommended: boolean;
};

File: apps/api/app/schemas/methods.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

MethodCategory = Literal["quantitative", "configurational"]
FieldType = Literal["string", "integer", "number", "boolean", "category", "date", "set_membership"]
ResultBlockType = Literal["summary", "table", "chart", "text", "download"]
AnalysisDifficulty = Literal["低", "中", "高"]


class MethodFieldConstraints(BaseModel):
    model_config = ConfigDict(extra="forbid")

    min: float | None = None
    max: float | None = None
    allowed_values: list[str] | None = None


class MethodFieldSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    label: str
    type: FieldType
    required: bool
    role: str
    allow_missing: bool = False
    constraints: MethodFieldConstraints | None = None
    description: str


class MethodInputSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec_version: str = "1.0"
    accepted_file_types: list[Literal["csv", "xlsx"]]
    min_sample_size: int | None = None
    fields: list[MethodFieldSpec]


class MethodResultBlockSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key: str
    type: ResultBlockType
    title: str
    required: bool


class MethodOutputSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    summary: str
    result_blocks: list[MethodResultBlockSpec]


class MethodSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    slug: str
    name: str
    category: MethodCategory
    description: str
    applicable_scenarios: list[str]
    data_type_label: str
    analysis_difficulty: AnalysisDifficulty
    sample_requirement: str
    is_recommended: bool = False


class MethodDetail(MethodSummary):
    input_spec_json: MethodInputSpec
    output_spec_json: MethodOutputSpec
    template_file_url: str | None = None
    sample_file_url: str | None = None
    prerequisites: list[str] = Field(default_factory=list)
    common_errors: list[str] = Field(default_factory=list)
    is_active: bool = True
    display_order: int = 0
    created_at: str
    updated_at: str
